Use side-stream in CPU to GPU copies in DDP

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50180 [WIP] Use side-stream in CPU to GPU copies in DDP**

Resolves the regression in
https://github.com/pytorch/pytorch/issues/49819 by adding copy over background
stream similar to scatter. We gate this with an env var which is by default true, but off for internal needs. 

Differential Revision: [D25818170](https://our.internmc.facebook.com/intern/diff/D25818170/)

rohan-varma committed:
[WIP] Use side-stream in CPU to GPU copies in DDP

Resolves the regression in
https://github.com/pytorch/pytorch/issues/49819 by adding copy over background
stream similar to scatter

Differential Revision: [D25818170](https://our.internmc.facebook.com/intern/diff/D25818170/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Use side-stream in CPU to GPU copies in DDP"


Resolves the regression in
https://github.com/pytorch/pytorch/issues/49819 by adding copy over background
stream similar to scatter

Differential Revision: [D25818170](https://our.internmc.facebook.com/intern/diff/D25818170/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Use side-stream in CPU to GPU copies in DDP"


Resolves the regression in
https://github.com/pytorch/pytorch/issues/49819 by adding copy over background
stream similar to scatter

Differential Revision: [D25818170](https://our.internmc.facebook.com/intern/diff/D25818170/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Use side-stream in CPU to GPU copies in DDP"


Resolves the regression in
https://github.com/pytorch/pytorch/issues/49819 by adding copy over background
stream similar to scatter

Differential Revision: [D25818170](https://our.internmc.facebook.com/intern/diff/D25818170/)

[ghstack-poisoned]

rohan-varma commented:
Using traces from the profiler on the test model in the test file (will be removed before commit), I'm not seeing too much of a difference before and after, wondering if this is expected? 

Before:
![Screen Shot 2021-01-06 at 11 14 02 PM](https://user-images.githubusercontent.com/8039770/103863489-cfa49e80-5075-11eb-9abf-ad89ea8125a2.png)

After:
![trace_cuda](https://user-images.githubusercontent.com/8039770/103863300-763c6f80-5075-11eb-9ff7-d0f989f5c075.png)

rohan-varma committed:
Update on "[WIP] Use side-stream in CPU to GPU copies in DDP"


Resolves the regression in
https://github.com/pytorch/pytorch/issues/49819 by adding copy over background
stream similar to scatter

Differential Revision: [D25818170](https://our.internmc.facebook.com/intern/diff/D25818170/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Use side-stream in CPU to GPU copies in DDP"


Resolves the regression in
https://github.com/pytorch/pytorch/issues/49819 by adding copy over background
stream similar to scatter

Differential Revision: [D25818170](https://our.internmc.facebook.com/intern/diff/D25818170/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] Use side-stream in CPU to GPU copies in DDP"


Resolves the regression in
https://github.com/pytorch/pytorch/issues/49819 by adding copy over background
stream similar to scatter

Differential Revision: [D25818170](https://our.internmc.facebook.com/intern/diff/D25818170/)

[ghstack-poisoned]

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -714,7 +715,20 @@ def _recursive_to(self, inputs, target_gpu):
-                return (obj.to(target_gpu), )
+                # Perform CPU -> GPU copies in a background stream. This code is
+                # motivated from similar logic in torch/nn/parallel/_functions.py
+                stream = _get_stream(target_gpu)
```
Shouldn't `torch.cuda.Stream()` give you a new stream from the pool? (https://github.com/pytorch/pytorch/blob/master/torch/csrc/cuda/Stream.cpp#L39)

rohan-varma replied:
@pritamdamania87 To simulate the loggic in `scatter` (see: https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/_functions.py#L85), I used `_get_stream` which is what scatter does. It looks like `_get_stream` calls into torch.cuda.Stream(), but it caches it for each device for re-use, instead of allocating a fresh stream everytime: https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/_functions.py#L112

pritamdamania87 replied:
Hmm I'm wondering what is the purpose of the caching in `_get_stream` since `torch.cuda.Stream()` also returns a stream from a pool and doesn't allocate a new one each time.

rohan-varma replied:
Are we guaranteed that torch.cuda.Stream() will give us the same stream everytime for a given device, though? I'm not sure if that's necessary here but the behavior with `_get_stream()` makes that the case. 

The original code of caching the streams in `_get_stream` is from a few years ago: https://github.com/pytorch/pytorch/pull/955

ngimel replied:
No, `torch.cuda.Stream()` won't give you the same stream every time. It is possible that at the time of #955 there was no stream pool, so caching the stream was a way to make sure a large number of streams was not created and destroyed.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -714,7 +715,20 @@ def _recursive_to(self, inputs, target_gpu):
+                    output = obj.to(target_gpu)
+                # synchronize with the copy stream
+                with torch.cuda.device(target_gpu):
+                    main_stream = torch.cuda.current_stream()
```
nit: current_stream instead of main_stream?

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -4389,3 +4389,26 @@ def test_scatter_object_list(self):
+        @require_backends_available({"gloo", "nccl"})
+        @skip_if_lt_x_gpu(2)
+        def test_ddp_cpu_copy(self):
+            # Test only for profiling purposes. Will be removed.
+            torch.cuda.set_device(self.rank)
+            dim = 10000
+            model = nn.Linear(dim, 1).to(self.rank)
+            model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[self.rank], output_device=self.rank)
+            inp = torch.randn(1000, dim)
+            with torch.autograd.profiler.profile(use_cuda=True) as p:
+                for i in range(6):
+                    with torch.autograd.profiler.record_function(f"##forward_{i}##"):
+
+                        out = model(inp)
+
+                    out.sum().backward()
+
+            if self.rank == 0:
+                p.export_chrome_trace("/tmp/trace_cuda_blocking.txt")
```
Can we add something like this to our benchmarking framework to catch regressions like this in the future? cc @bitfort

rohan-varma committed:
Update on "[WIP] Use side-stream in CPU to GPU copies in DDP"


Resolves the regression in
https://github.com/pytorch/pytorch/issues/49819 by adding copy over background
stream similar to scatter

Differential Revision: [D25818170](https://our.internmc.facebook.com/intern/diff/D25818170/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -4389,3 +4389,26 @@ def test_scatter_object_list(self):
+        @require_backend({"gloo", "nccl"})
+        @require_backends_available({"gloo", "nccl"})
+        @skip_if_lt_x_gpu(2)
+        def test_ddp_cpu_copy(self):
```
will rm before landing

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -714,7 +715,20 @@ def _recursive_to(self, inputs, target_gpu):
+                with torch.cuda.device(target_gpu):
+                    current_stream = torch.cuda.current_stream()
+                    # Sync the current stream with the copy stream
+                    current_stream.wait_stream(stream)
```
isn't the copy operator already doing this in c++ code?

https://github.com/pytorch/pytorch/blob/7b21c6be67bc8669e7b93e641a363bbb84b4856d/aten/src/ATen/native/cuda/Copy.cu#L90

rohan-varma replied:
Good point, although I wanted to keep this change as similar as `scatter_gather` code as possible to prevent regressions/errors: https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/_functions.py#L102. We can probably evaluate whether to change this in both callsites in a follow-up change?

mrshenli reviewed (approved):
LGTM! Just have a minor question on stream sync.

rohan-varma committed:
Update on "[WIP] Use side-stream in CPU to GPU copies in DDP"


Resolves the regression in
https://github.com/pytorch/pytorch/issues/49819 by adding copy over background
stream similar to scatter

Differential Revision: [D25818170](https://our.internmc.facebook.com/intern/diff/D25818170/)

[ghstack-poisoned]

mrshenli reviewed (approved):
Stamp on the env var change.

rohan-varma committed:
Update on "[WIP] Use side-stream in CPU to GPU copies in DDP"


Resolves the regression in
https://github.com/pytorch/pytorch/issues/49819 by adding copy over background
stream similar to scatter. We gate this with an env var which is by default true, but off for internal needs. 

Differential Revision: [D25818170](https://our.internmc.facebook.com/intern/diff/D25818170/)

[ghstack-poisoned]
